### PR TITLE
Cookies

### LINKED ISSUE #53: Add Cookies API
There should be an API to access cookies.

### PR DESCRIPTION
**Pull request by JoelEinbinder**
Closes #53

### CODE REVIEW

File: docs/api.md [hunk 1]
```diff
@@ -315,6 +318,30 @@ Shortcut for [page.mainFrame().click(selector[, options])](#frameclickselector-o
...
+  - `expires` <[number]> Unix time in seconds.
+  - `httpOnly` <[boolean]>
+  - `secure` <[boolean]>
+  - `sameSite` <[string]> `"Strict"` or `"Lax"`.
+
+If no URLs are specified, this method returns cookies related to the current page.
+If URLs are specified, only cookies for those URLs are returned.
+
+#### page.deleteCookie(...cookies)
+- `...cookies` <[Object]>
```
Comment (aslushnikov) [inline docs/api.md]: here and below: <[Array]<[Object]>>

Comment (JoelEinbinder) [reply docs/api.md]: I don't like putting array when we have a spread operator. The user doesn't pass in an array.

**Comment by JoelEinbinder** [comment]
An update to the cookie protocol is landing in r495262

https://chromium-review.googlesource.com/c/617750
also https://chromium-review.googlesource.com/c/614721 landed recently.

File: docs/api.md [hunk 2]
```diff
@@ -315,6 +318,30 @@ Shortcut for [page.mainFrame().click(selector[, options])](#frameclickselector-o
...
+  - `name` <[string]>
+  - `value` <[string]>
+  - `domain` <[string]>
+  - `path` <[string]>
+  - `expires` <[number]> Unix time in seconds.
+  - `httpOnly` <[boolean]>
+  - `secure` <[boolean]>
+  - `sameSite` <[string]> `"Strict"` or `"Lax"`.
+
+If no URLs are specified, this method returns cookies related to the current page.
```
Comment (pavelfeldman) [inline docs/api.md]: I don't think you should inherit this poor DTP semantics. There is no real 'related to the page', there are cookies matching domains, paths and names. I would suggest to add optional name, domain and path parameters that would match the cookies. Your underlying implementation would be fetching everything using `Network.getAllCookies`.

File: docs/api.md [hunk 3]
```diff
@@ -315,6 +318,30 @@ Shortcut for [page.mainFrame().click(selector[, options])](#frameclickselector-o
...
+If URLs are specified, only cookies for those URLs are returned.
+
+#### page.deleteCookie(...cookies)
+- `...cookies` <...[Object]>
+  - `name` <[string]> **required**
+  - `url` <[string]>
```
Comment (pavelfeldman) [inline docs/api.md:362]: Either url or domain should be required for scoping. Also, this method will delete multiple cookies for each pattern, so might be worth calling method `deleteCookies` and parameters `patterns`

File: docs/api.md [hunk 4]
```diff
@@ -611,6 +638,19 @@ Shortcut for [`keyboard.down`](#keyboarddownkey-options) and [`keyboard.up`](#ke
 - `html` <[string]> HTML markup to assign to the page.
 - returns: <[Promise]>
 
+#### page.setCookie(...cookies)
+- `...cookies` <...[Object]>
+  - `name` <[string]> **required**
+  - `value` <[string]> **required**
+  - `url` <[string]>
```
Comment (pavelfeldman) [inline docs/api.md:686]: either url or domain required.

Comment (JoelEinbinder) [reply docs/api.md:686]: It uses the page url if none is specified.

File: docs/api.md (hunk 2 above)
Comment (JoelEinbinder) [reply docs/api.md]: I'd agree if the method was browser.setCookie. But page.setCookie should act like I set a cookie for the page.

Comment (pavelfeldman) [reply docs/api.md]: I don't think you know what `Network.getCookies` without `urls` does...

File: docs/api.md (hunk 4 above)
Comment (pavelfeldman) [reply docs/api.md:686]: Cookies are not associated with the page, they are associated with the document (frame).

File: docs/api.md (hunk 2 above)
Comment (JoelEinbinder) [reply docs/api.md]: Ok now I don't think I do either.

Comment (JoelEinbinder) [reply docs/api.md]: I want to be able to do
```js
page.setCookie({name, value});
page.deleteCookie({name, value});
```

Comment (JoelEinbinder) [reply docs/api.md]: Disregard that, I read this comment as being on setCookie.

Comment (pavelfeldman) [reply docs/api.md]: (I think your understanding of 'related to the page' is different from the protocol one. I'm fine with your understanding, but it needs to be clarified and implementation needs to be changed. Also, you might want to pull this into the frame and alias in page)

**Review by MiniCodeMonkey — COMMENTED** [review]
I'm super excited for this feature!

File: lib/Page.js [hunk 5]
```diff
@@ -155,6 +155,47 @@ class Page extends EventEmitter {
...
+      urls: urls.length ? urls : [this.url()]
+    })).cookies;
+  }
+
+  /**
+   * @param {Array<Network.CookieParam>} cookies
+   */
+  async deleteCookie(...cookies) {
+    for (const cookie of cookies) {
+      const item = {};
```
Comment (aslushnikov) [inline lib/Page.js]:
```js
const item = Object.assign({}, cookie);
```

File: test/test.js [hunk 6]
```diff
@@ -1920,6 +1920,162 @@ describe('Page', function() {
...
+        value: '1'
+      }, {
+        name: 'cookie2',
+        value: '2'
+      }, {
+        name: 'cookie3',
+        value: '3'
+      });
+      expect(await page.evaluate('document.cookie')).toBe('cookie1=1; cookie2=2; cookie3=3');
+      page.deleteCookie({name: 'cookie2'});
```
Comment (aslushnikov) [inline test/test.js]: technically, you'd want to `await` it here

File: lib/Page.js [hunk 7]
```diff
@@ -155,6 +155,47 @@ class Page extends EventEmitter {
...
+      Object.assign(item, cookie);
+      const pageURL = this.url();
```
Comment (aslushnikov) [inline lib/Page.js]: nit: bring pageURL out of the loop

**Review by aslushnikov — APPROVED** [review]
lgtm % nits

**Pull request merged**